crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -880,6 +881,36 @@ where
 
         self.current_match_case.as_mut().unwrap().index += 1;
     }
+
+    fn visit_except_handler(&mut self, except_handler: &'ast ast::ExceptHandler) {
+        let ast::ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler {
+            name,
+            type_,
+            body,
+            range: _,
+        }) = except_handler;
+        if let Some(type_) = type_ {
+            if let Some(name) = name {
+                self.scopes_by_expression
+                    .insert(name.into(), self.current_scope());
+                self.current_ast_ids().record_expression(name);
```

Comment:
What's the reason that we need to treat the name as an expression?

Reply:
This is the same issue I highlighted in https://github.com/astral-sh/ruff/pull/13267/files#r1746998476, more or less. `HasScopedId::scoped_ast_id()` needs to be called in `infer_except_handler_definition()` on an `ast::Identifier` node in order to store the type associated with this definition. This means that `HasScopedId::scoped_ast_id()` must be implemented for `ast::Identifier`; but `HasScopedId::scoped_ast_id()` won't work for the `ast::Identifier` nodes here unless we also store them in the `AstIds::expressions_map`. Otherwise we end up with this crash:

```
---- types::infer::tests::except_handler stdout ----
thread 'types::infer::tests::except_handler' panicked at crates/red_knot_python_semantic/src/semantic_index/ast_ids.rs:37:29:
no entry found for key
```

I suppose we could store the except-handler IDs in a different map? `AstIds::identifiers_map`? And look up the `Identifier` node in that map from the `HasScopedId::scoped_ast_id` implementation for `ast::Identifier`? It seems a little wasteful to create a separate `HashMap` just for this purpose, though, since the vast majority of definitions have `Expr` nodes as their targets -- except-handler definitions are somewhat unique here?

Reply:
What's the reason that we need to store the type of the symbol? Isn't it enough to just store the type of the definition?

Reply:
I see what you mean. Sorry, I guess I'm still getting to grips with some of the semantic-index code. I pushed https://github.com/astral-sh/ruff/pull/13267/commits/7b507e8ddc0d859f2a4b4b47bde9c7244fb195d9, and all tests seem to pass, at least for now.

Reply:
> Edit: Do we need to implement `HasTy` for `ExceptHandler` and explicitly call it in the corpus test?

I'm not sure it would be correct for `ExceptHandler` to implement `HasTy`. What would be the type of `ExceptHandler`s like these? In the first one, the `name` field is set to `None`; in the second one, `name` and `type_` are both `None`:

```py
try:
    x
except NameError:
    pass

try:
    y
except:
    pass
```

Reply:
Good point. I think the way it would be done is to call `.ty` on the handled exception expression? Interested to hear what @carljm thinks

Reply:
I didn't see this code anymore when I reviewed the current version of the PR, so I think it must have already been removed? I don't actually think we should make the assumption anywhere that assignment targets must be Name nodes; that's not true for imports either.

Regarding `HasTy`, I generally think it's not important that we implement it for all non-expression definition nodes, if it's not self-evident that all nodes of that kind will have a clear and obvious type. I think the main problem here is bare `except` -- if it weren't for that, we _could_ implement `HasTy` for all `ExceptHandler` nodes, and it could be the handled-exceptions union that we get from analyzing the handled-exceptions (the fact that the exception isn't captured in a name doesn't stop us from implementing `HasTy`).

But I also don't think it's _important_ that we implement `HasTy` for except-handler nodes, and since the bare `except` problem is a little thorny, I would probably just not bother.